runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -1196,42 +1196,62 @@ end
 ---             - Exclusive severity to consider. Overrides {severity_limit}
 ---         - {severity_limit}: (DiagnosticSeverity)
 ---             - Limit severity of diagnostics found. E.g. "Warning" means { "Error", "Warning" } will be valid.
+---         - {workspace}: (boolean, default false)
+---             - Set the list with workspace diagnostics
 function M.set_loclist(opts)
   opts = opts or {}
 
   local open_loclist = if_nil(opts.open_loclist, true)
 
-  local bufnr = vim.api.nvim_get_current_buf()
-  local buffer_diags = M.get(bufnr, opts.client_id)
-
-  if opts.severity then
-    buffer_diags = filter_to_severity_limit(opts.severity, buffer_diags)
-  elseif opts.severity_limit then
-    buffer_diags = filter_by_severity_limit(opts.severity_limit, buffer_diags)
-  end
+  local current_bufnr = vim.api.nvim_get_current_buf()
+  local diags = opts.workspace and M.get_all() or {
+    [current_bufnr] = M.get(current_bufnr, opts.client_id)
+  }
 
   local items = {}
-  local insert_diag = function(diag)
+  local insert_diag = function(bufnr, diag)
     local pos = diag.range.start
     local row = pos.line
-    local col = util.character_offset(bufnr, row, pos.character)
 
-    local line = (api.nvim_buf_get_lines(bufnr, row, row + 1, false) or {""})[1]
+    local line, col, text
+    if api.nvim_buf_is_loaded(bufnr) then
+      line = (api.nvim_buf_get_lines(bufnr, row, row + 1, false) or {""})[1]
+      col = util.character_offset(bufnr, row, pos.character) or 0
+      text = line .. " | " .. diag.message
+    else
+      col = pos.character
+      text = diag.message
+    end
+
 
     table.insert(items, {
       bufnr = bufnr,
       lnum = row + 1,
       col = col + 1,
-      text = line .. " | " .. diag.message,
+      text = text,
       type = loclist_type_map[diag.severity or DiagnosticSeverity.Error] or 'E',
     })
   end
 
-  for _, diag in ipairs(buffer_diags) do
-    insert_diag(diag)
+  for bufnr, diagnostic in pairs(diags) do
+    if opts.severity then
+      diagnostic = filter_to_severity_limit(opts.severity, diagnostic)
+    elseif opts.severity_limit then
+      diagnostic = filter_by_severity_limit(opts.severity_limit, diagnostic)
+    end
+
+    for _, diag in ipairs(diagnostic) do
+      insert_diag(bufnr, diag)
+    end
   end
 
-  table.sort(items, function(a, b) return a.lnum < b.lnum end)
+  table.sort(items, function(a, b)
+    if a.bufnr == b.bufnr then
+      return a.lnum < b.lnum
+    else
+      return a.bufnr < b.bufnr
+    end
+  end)
 
   util.set_loclist(items)
```

Comment:
Do you think it would make sense to split `set_loclist` into a function that returns the diagnostics in loclist/qflist format, so that users could also easily populate the quickfix list instead of the locationlist with the result?

The quickfix list might make more sense for diagnostics that are not specific to the buffer in the current window.

Reply:
it does to me, though what should the name of this function be?